File: archive/pulumi-py/Pr5831/lib/infrastructure/api_gateway.py
"""
API Gateway module for HTTP API management.

This module creates and configures API Gateway HTTP APIs with proper
Lambda integration, throttling, and CORS settings.
"""

import json
from typing import Dict

import pulumi_aws as aws
from pulumi import Output

from .aws_provider import AWSProviderManager
from .config import ServerlessProcessorConfig
from .lambda_functions import LambdaStack


class APIGatewayStack:
    """
    Manages API Gateway for the serverless processor.
    
    Creates HTTP API Gateway with Lambda integration, proper permissions,
    throttling, and CORS configuration.
    """
    
    def __init__(
        self,
        config: ServerlessProcessorConfig,
        provider_manager: AWSProviderManager,
        lambda_stack: LambdaStack
    ):
        """
        Initialize the API Gateway stack.
        
        Args:
            config: ServerlessProcessorConfig instance
            provider_manager: AWSProviderManager instance
            lambda_stack: LambdaStack instance for Lambda integration
        """
        self.config = config
        self.provider_manager = provider_manager
        self.lambda_stack = lambda_stack
        self.api = None
        self.stage = None
        self.integration = None
        self.route = None
        
        self._create_api()
    
    def _create_api(self):
        """Create API Gateway HTTP API."""
        api_name = self.config.get_resource_name('api')
        
        self.api = aws.apigatewayv2.Api(
            'http-api',
            name=api_name,
            protocol_type='HTTP',
            cors_configuration=aws.apigatewayv2.ApiCorsConfigurationArgs(
                allow_origins=['*'],
                allow_methods=['POST', 'OPTIONS'],
                allow_headers=['Content-Type', 'X-Request-ID', 'Authorization'],
                max_age=300
            ),
            tags={
                **self.config.get_common_tags(),
                'Name': api_name
            },
            opts=self.provider_manager.get_resource_options()
        )
        
        processor_function = self.lambda_stack.get_function('processor')
        
        self.integration = aws.apigatewayv2.Integration(
            'lambda-integration',
            api_id=self.api.id,
            integration_type='AWS_PROXY',
            integration_uri=self.lambda_stack.get_function_invoke_arn('processor'),
            integration_method='POST',
            payload_format_version='2.0',
            timeout_milliseconds=self.config.lambda_timeout * 1000,
            opts=self.provider_manager.get_resource_options()
        )
        
        self.route = aws.apigatewayv2.Route(
            'process-route',
            api_id=self.api.id,
            route_key='POST /process',
            target=Output.concat('integrations/', self.integration.id),
            opts=self.provider_manager.get_resource_options()
        )
        
        log_group_name = f'/aws/apigatewayv2/{api_name}'
        
        api_log_group = aws.cloudwatch.LogGroup(
            'api-log-group',
            name=log_group_name,
            retention_in_days=self.config.log_retention_days,
            tags={
                **self.config.get_common_tags(),
                'Name': log_group_name
            },
            opts=self.provider_manager.get_resource_options()
        )
        
        self.stage = aws.apigatewayv2.Stage(
            'api-stage',
            api_id=self.api.id,
            name=self.config.api_stage_name,
            auto_deploy=True,
            access_log_settings=aws.apigatewayv2.StageAccessLogSettingsArgs(
                destination_arn=api_log_group.arn,
                format=json.dumps({
                    'requestId': '$context.requestId',
                    'ip': '$context.identity.sourceIp',
                    'requestTime': '$context.requestTime',
                    'httpMethod': '$context.httpMethod',
                    'routeKey': '$context.routeKey',
                    'status': '$context.status',
                    'protocol': '$context.protocol',
                    'responseLength': '$context.responseLength',
                    'error': '$context.error.message',
                    'integrationError': '$context.integration.error'
                })
            ),
            default_route_settings=aws.apigatewayv2.StageDefaultRouteSettingsArgs(
                throttling_burst_limit=self.config.api_throttle_burst_limit,
                throttling_rate_limit=self.config.api_throttle_rate_limit
            ),
            tags={
                **self.config.get_common_tags(),
                'Name': f'{api_name}-{self.config.api_stage_name}'
            },
            opts=self.provider_manager.get_resource_options(depends_on=[self.route, api_log_group])
        )
        
        aws.lambda_.Permission(
            'api-lambda-permission',
            action='lambda:InvokeFunction',
            function=processor_function.name,
            principal='apigateway.amazonaws.com',
            source_arn=Output.all(self.api.execution_arn, self.stage.name).apply(
                lambda args: f'{args[0]}/{args[1]}/POST/process'
            ),
            opts=self.provider_manager.get_resource_options()
        )
    
    def get_api_id(self) -> Output[str]:
        """
        Get API Gateway ID.
        
        Returns:
            API ID as Output
        """
        return self.api.id
    
    def get_api_endpoint(self) -> Output[str]:
        """
        Get API Gateway endpoint URL.
        
        Returns:
            API endpoint URL as Output
        """
        return self.api.api_endpoint
    
    def get_api_url(self) -> Output[str]:
        """
        Get full API URL with stage and path.
        
        Returns:
            Full API URL as Output
        """
        return Output.all(self.api.api_endpoint, self.stage.name).apply(
            lambda args: f'{args[0]}/{args[1]}/process'
        )

File: archive/pulumi-ts/Pr6619/lib/components/api-gateway.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';
import { EnvironmentConfig, TagsConfig } from '../types';

export interface ApiGatewayComponentArgs {
  environmentSuffix: string;
  envConfig: EnvironmentConfig;
  tags: TagsConfig;
  lambdaFunctionArn: pulumi.Output<string>;
  lambdaFunctionName: pulumi.Output<string>;
}

export class ApiGatewayComponent extends pulumi.ComponentResource {
  public readonly api: aws.apigatewayv2.Api;
  public readonly stage: aws.apigatewayv2.Stage;
  public readonly wafWebAcl?: aws.wafv2.WebAcl;
  public readonly wafAssociation?: aws.wafv2.WebAclAssociation;
  public readonly logGroup: aws.cloudwatch.LogGroup;
  public readonly apiEndpoint: pulumi.Output<string>;
  public readonly apiArn: pulumi.Output<string>;

  constructor(
    name: string,
    args: ApiGatewayComponentArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:apigateway:ApiGatewayComponent', name, {}, opts);

    const {
      environmentSuffix,
      envConfig,
      tags,
      lambdaFunctionArn,
      lambdaFunctionName,
    } = args;

    // Create API Gateway HTTP API
    this.api = new aws.apigatewayv2.Api(
      `payment-api-${environmentSuffix}`,
      {
        name: `payment-api-${environmentSuffix}`,
        protocolType: 'HTTP',
        description: `Payment processing API for ${envConfig.environment} environment`,
        corsConfiguration: {
          allowOrigins: ['*'],
          allowMethods: ['POST', 'GET', 'OPTIONS'],
          allowHeaders: ['Content-Type', 'Authorization'],
          maxAge: 300,
        },
        tags: {
          ...tags,
          Name: `payment-api-${environmentSuffix}`,
          CustomDomain: envConfig.customDomain,
        },
      },
      { parent: this }
    );

    // Create Lambda integration
    const integration = new aws.apigatewayv2.Integration(
      `payment-api-integration-${environmentSuffix}`,
      {
        apiId: this.api.id,
        integrationType: 'AWS_PROXY',
        integrationUri: lambdaFunctionArn,
        integrationMethod: 'POST',
        payloadFormatVersion: '2.0',
        timeoutMilliseconds: 30000,
      },
      { parent: this }
    );

    // Create routes
    new aws.apigatewayv2.Route(
      `payment-post-route-${environmentSuffix}`,
      {
        apiId: this.api.id,
        routeKey: 'POST /payment',
        target: pulumi.interpolate`integrations/${integration.id}`,
      },
      { parent: this }
    );

    new aws.apigatewayv2.Route(
      `payment-get-route-${environmentSuffix}`,
      {
        apiId: this.api.id,
        routeKey: 'GET /payment/{id}',
        target: pulumi.interpolate`integrations/${integration.id}`,
      },
      { parent: this }
    );

    // Create CloudWatch log group for API Gateway
    this.logGroup = new aws.cloudwatch.LogGroup(
      `payment-api-logs-${environmentSuffix}`,
      {
        name: `/aws/apigateway/payment-api-${environmentSuffix}`,
        retentionInDays: envConfig.logRetentionDays,
        tags: {
          ...tags,
        },
      },
      { parent: this }
    );

    // Create stage with auto-deploy
    this.stage = new aws.apigatewayv2.Stage(
      `payment-api-stage-${environmentSuffix}`,
      {
        apiId: this.api.id,
        name: envConfig.environment,
        autoDeploy: true,
        accessLogSettings: {
          destinationArn: this.logGroup.arn,
          format: JSON.stringify({
            requestId: '$context.requestId',
            ip: '$context.identity.sourceIp',
            requestTime: '$context.requestTime',
            httpMethod: '$context.httpMethod',
            routeKey: '$context.routeKey',
            status: '$context.status',
            protocol: '$context.protocol',
            responseLength: '$context.responseLength',
          }),
        },
        tags: {
          ...tags,
        },
      },
      { parent: this }
    );

    // Grant API Gateway permission to invoke Lambda
    new aws.lambda.Permission(
      `payment-lambda-apigw-permission-${environmentSuffix}`,
      {
        action: 'lambda:InvokeFunction',
        function: lambdaFunctionName,
        principal: 'apigateway.amazonaws.com',
        sourceArn: pulumi.interpolate`${this.api.executionArn}/*/*`,
      },
      { parent: this }
    );

    // Create AWS WAF WebACL for prod environment only
    if (envConfig.enableWaf) {
      this.wafWebAcl = new aws.wafv2.WebAcl(
        `payment-waf-${environmentSuffix}`,
        {
          name: `payment-waf-${environmentSuffix}`,
          scope: 'REGIONAL',
          description: `WAF for payment API in ${envConfig.environment}`,
          defaultAction: {
            allow: {},
          },
          rules: [
            {
              name: 'RateLimitRule',
              priority: 1,
              action: {
                block: {},
              },
              statement: {
                rateBasedStatement: {
                  limit: 2000,
                  aggregateKeyType: 'IP',
                },
              },
              visibilityConfig: {
                cloudwatchMetricsEnabled: true,
                metricName: 'RateLimitRule',
                sampledRequestsEnabled: true,
              },
            },
            {
              name: 'AWSManagedRulesCommonRuleSet',
              priority: 2,
              overrideAction: {
                none: {},
              },
              statement: {
                managedRuleGroupStatement: {
                  vendorName: 'AWS',
                  name: 'AWSManagedRulesCommonRuleSet',
                },
              },
              visibilityConfig: {
                cloudwatchMetricsEnabled: true,
                metricName: 'AWSManagedRulesCommonRuleSet',
                sampledRequestsEnabled: true,
              },
            },
          ],
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: `payment-waf-${environmentSuffix}`,
            sampledRequestsEnabled: true,
          },
          tags: {
            ...tags,
          },
        },
        { parent: this }
      );

      // Associate WAF with API Gateway
      this.wafAssociation = new aws.wafv2.WebAclAssociation(
        `payment-waf-association-${environmentSuffix}`,
        {
          resourceArn: this.stage.arn,
          webAclArn: this.wafWebAcl.arn,
        },
        { parent: this }
      );
    }

    this.apiEndpoint = pulumi.interpolate`${this.api.apiEndpoint}/${this.stage.name}`;
    this.apiArn = this.api.arn;

    this.registerOutputs({
      apiEndpoint: this.apiEndpoint,
      apiArn: this.apiArn,
    });
  }
}
